contributors/design-proposals/controller_history.md

```diff
@@ -0,0 +1,447 @@
+    to ensure that only non-live version snapshots are collected.
+1. Controllers may also use the `OwnerReferences` field of the AnyState to 
+record all Objects that are generated from target Object state version 
+represented by the AnyState as its owners.
+   - A revision is considered to be live while any generated Object that owns 
+   it is live.
+   - This method allows for the implementation of generic garbage collection.
+   - The primary drawback with this method is that the book keeping is complex, 
+   and deciding if a generated Object corresponds to a particular revision 
+   will require testing each Object for membership in the `OwnerReferences` 
+   of all AnyStates.
+   
+At the cost of the complexity of implementing both labeling and ownership, 
+controllers may use a combination of both approaches to mitigate the 
+deficiencies of each.
+
+### Version Equivalence
+When the target Object state of a specification type Object is revised, we wish 
+to minimize the number of mutations to generated Objects as the controller seeks
+to conform the system to its target state. That is, if a generated Object 
+already conforms to the revised target Object state, it is imperative that we 
+do not mutate it. 
+
+Failure to implement this correctly could result in the simultaneous rolling 
+restart of every Pod in every StatefulSet and DaemonSet in the system when 
+additions are made to PodTemplateSpec during a master upgrade. It is therefore 
+necessary to determine if the current target Object state is equivalent to a 
+prior version. 
+
+Since we [track the version of](#version-tracking) of generated Objects, this 
+reduces to deciding if the version of the target Object state associated with 
+the generated Object is equivalent to the current target Object state.
+Even though  [hashing](#hashing) is used to generate the `.Name` of the 
+AnyStates used to encapsulate versions of the target Object state, as we do 
+not require cryptographically strong collision resistance, and given we use a 
+[collision resolution](#collision-resolution) technique, we can't use the 
+[generated names](#unique-name-generation) of AnyStates to decide equality.
+
+We propose that two AnyStates can be considered equal if their `.Data` is 
+equivalent, but that it is not sufficient to compare the serialized 
+representation of the their `.Data`. Consider that the addition of new fields 
+to the Objects that represent the target Object state may cause the serialized 
+representation of those Objects to be unequal even when they are semantically 
+equivalent. We propose two methods that controllers may use to decide the 
+equivalence of two versions of their target Object state.
+
+1. The controller may deserialize the values of the AnyStates representing their 
+target Object state and perform a deep, semantic equality test. Here all 
+differences that do not constitute a mutation to the target Object state are 
+disregarded during the equivalence test.
```

Comment:
Deserialize each object and write a long, Java style, object equivalence comparison. Every time a new field gets added, update the imperative code that performs the comparison. As indicated below, this might get us off the ground but should probably not be the long term solution. Strategic merge patch is the way to go in the long run.